chiselFrontend/src/main/scala/chisel3/core/Bits.scala

```diff
@@ -67,6 +90,7 @@ sealed abstract class Bits(dirArg: Direction, width: Width, override val litArg:
     if (isLit()) {
       Bool(((litValue() >> x.toInt) & 1) == 1)
     } else {
+      Binding.checkSynthesizable(this, s"'this' ($this)")
```

Comment:
It is not a violation of DRY to verify that invariants/constraints are satisfied between program steps. It would be a violation if you needlessly duplicated code. However, the Chisel and FIRRTL data structures and machinery are so different that it isn't trivial to reuse code. (That said, it would be really good if they shared width propagation assumptions since those seem to be getting out of sync)

Note, the underlying problem is that Chisel was emitting majorly malformed FIRRTL ASTs with references to undefined things. In a lot of cases this means that you forgot a Wire or Reg wrapper around some Data. However, this can also occur when you improperly reference an internal node from a different Module. (At best, it isn't defined in the current module, it is defined in the current Module but with an incompatible type, and at worst, it is defined and is of a compatible type.) In the current scheme, FIRRTL cannot efficiently differentiate between the two.

You could tag each Reference with the module that owns it; however, at that point you might as well just error since the frontend would have to have looked up the Reference owners anyway.....

Related to the latter case (I think, I forget the exact particulars on what was happening internally), you get some really 'fun' issues if you have an IO Bundle that instantiates logic in its constructor (e.g. by accidentally using Vec.fill....).